added dylib flag, builds sokol_clib as a dynamic library

remzisenel opened the pull request:
This change will enable building the library as a dynamic library, which in turn enables hot reloading of sokol.

floooh commented on build.zig:
```diff
@@ -152,11 +155,17 @@ pub fn buildLibSokol(b: *Build, options: LibSokolOptions) !*Build.Step.Compile {
+            true => .dynamic,
+            else => .static,
+        },
+        .root_module = b.addModule("sokol_clib", .{
+            .target = options.target,
+            .optimize = options.optimize,
+            .link_libc = true,
+        }),
```
This change looks like it's mixing the 'old' and 'new' way, since on one hand, the `root_module` approach is used, but on the other hand, the C source files and system link libraries are added to the `lib` build step instead of to the module.

Please move the module creation out of the `b.addLibrary()` and rename the module name to not clash with the library name, e.g. smth like this:

```zig
const mod = b.addModule("mod_sokol_clib", .{
            .target = options.target,
            .optimize = options.optimize,
            .link_libc = true,
});

...then change all the `addCSourceFile` and `linkSystemLibrary` calls to happen on the module object.

Finally move the creation of the library build step to the end of the function (and I would also change the switch to an if, since it tests on a boolean:

```zig
const lib = b.addLibrary(.{
    .name = "sokol_clib",
    .root_module = mod,
    .linkage = if (options.dylib) .dynamic else .static,
});
```

floooh commented on build.zig:
```diff
@@ -131,6 +133,7 @@ pub const LibSokolOptions = struct {
     backend: SokolBackend = .auto,
     use_egl: bool = false,
     use_x11: bool = true,
+    dylib: bool = false,
```
I wonder if people outside the Apple ecosystem know what a `dylib` is, but I don't have an immediately better suggestion (except maybe `dynamic_linkage` because it matches the Zig build system lingo).

floooh commented on build.zig:
```diff
@@ -72,6 +72,7 @@ pub fn build(b: *Build) !void {
     const opt_dont_link_system_libs = b.option(bool, "dont_link_system_libs", "Do not link system libraries required by sokol (default: false)") orelse false;
     const opt_sokol_imgui_cprefix = b.option([]const u8, "sokol_imgui_cprefix", "Override Dear ImGui C bindings prefix for sokol_imgui.h (see SOKOL_IMGUI_CPREFIX)");
     const opt_cimgui_header_path = b.option([]const u8, "cimgui_header_path", "Override the Dear ImGui C bindings header name (default: cimgui.h)");
+    const opt_dylib = b.option(bool, "dylib", "Builds sokol_clib as dylib") orelse false;
```
See other comment about naming.

floooh reviewed (changes requested):
Thanks for the PR! I added a couple of code review comments, please fix at least the lib- vs module-confusion when adding source files and system link libraries. I'm flexible about the 'dylib' name, but would be nice if we can find something more 'cross-platform'.

remzisenel commented:
These changes were a bit heavier than I anticipated, I'm not fully familiar with Emscripten SDK so I tried my best not to meddle with anything unrelated. There are two items I would appreciate your attention to:

- isPlatform() conditionals (and other isPlatform() conditionals within resolveSokolBackend()) were using rootModuleTarget(), I introduced a variable mod_target, directly copies options.target.result. I think the behavior should be the same as providing the target to addStaticLibrary and retrieving it through rootModuleTarget().
- I moved emSdkSetupStep(), so it is closer to library artifact definition (which was moved to the end of the function as well).

Let me know what you think, thanks!

floooh commented:
Thanks for the changes! I'll give it a whirl on my Mac!

floooh reviewed (approved):
Tested and all looking good. Merging now. Thanks!

floooh commented:
And merged. Looking into the dcimgui PR now.
